File: Backend/src/Services/Inventory/Softaxis.Inventory.Domain/Entities/StockMovement.cs
namespace Softaxis.Inventory.Domain.Entities;

public sealed class StockMovement
{
    private StockMovement() { }

    public StockMovement(
        Guid productId, string movementType, decimal quantity,
        decimal unitCost, string? reference, string? notes, string? warehouseId)
    {
        Id           = Guid.NewGuid();
        ProductId    = productId;
        MovementType = movementType; // Receipt | Sale | Adjustment | Transfer | WriteOff | Return
        Quantity     = quantity;
        UnitCost     = unitCost;
        Reference    = reference?.Trim();
        Notes        = notes?.Trim();
        WarehouseId  = string.IsNullOrWhiteSpace(warehouseId) ? null : Guid.Parse(warehouseId);
        MovedAt      = DateTime.UtcNow;
        CreatedAt    = DateTime.UtcNow;
    }

    public Guid     Id           { get; private set; }
    public Guid     ProductId    { get; private set; }
    public string   MovementType { get; private set; } = string.Empty;
    public decimal  Quantity     { get; private set; }  // positive = in, negative = out
    public decimal  UnitCost     { get; private set; }
    public string?  Reference    { get; private set; }  // PO number, SO number, etc.
    public string?  Notes        { get; private set; }
    public Guid?    WarehouseId  { get; private set; }
    public DateTime MovedAt      { get; private set; }
    public DateTime CreatedAt    { get; private set; }
    public bool     IsDeleted    { get; private set; }

    // Navigation
    public Product?   Product   { get; private set; }
    public Warehouse? Warehouse { get; private set; }

    public void Delete() { IsDeleted = true; }
}

File: BackendNode/src/modules/inventory/models/StockMovement.js
const mongoose = require('mongoose');

const schema = new mongoose.Schema({
  tenantId:    { type: String, required: true, index: true },
  productId:   { type: String, required: true },
  productName: String,
  warehouseId: String,
  warehouseName:String,
  type:        { type: String, enum: ['receipt', 'write-off', 'adjustment', 'count-correction', 'transfer-in', 'transfer-out'], required: true },
  quantity:    { type: Number, required: true },
  unitCost:    { type: Number, default: 0 },
  reference:   String,
  notes:       String,
  createdAt:   { type: Date, default: Date.now },
}, { versionKey: false });

schema.index({ tenantId: 1, productId: 1 });

module.exports = mongoose.model('StockMovement', schema);
